essentials/src/historical_subscription.rs

```diff
@@ -0,0 +1,183 @@
+	types::BlockNumber,
+	utils::RetryOptions,
+};
+use async_trait::async_trait;
+use futures::future;
+use log::{debug, error, info};
+use polkadot_introspector_priority_channel::{channel, Sender};
+use tokio::{
+	sync::broadcast::Sender as BroadcastSender,
+	time::{interval_at, Duration},
+};
+
+pub struct HistoricalSubscription {
+	urls: Vec<String>,
+	from_block_number: BlockNumber,
+	to_block_number: BlockNumber,
+	/// One sender per consumer per URL.
+	consumers: Vec<Vec<Sender<ChainSubscriptionEvent>>>,
+	retry: RetryOptions,
+}
+
+#[async_trait]
+impl EventStream for HistoricalSubscription {
+	type Event = ChainSubscriptionEvent;
+
+	/// Create a new consumer of events. Returns consumer initialization data.
+	fn create_consumer(&mut self) -> EventConsumerInit<Self::Event> {
+		let mut update_channels = Vec::new();
+
+		// Create per ws update channels.
+		for _ in 0..self.urls.len() {
+			update_channels.push(channel(MAX_MSG_QUEUE_SIZE));
+		}
+
+		let (update_tx, update_channels) = update_channels.into_iter().unzip();
+
+		// Keep per url update senders for this consumer.
+		self.consumers.push(update_tx);
+
+		EventConsumerInit::new(update_channels)
+	}
+
+	async fn run(
+		self,
+		tasks: Vec<tokio::task::JoinHandle<()>>,
+		shutdown_tx: BroadcastSender<()>,
+		shutdown_future: tokio::task::JoinHandle<()>,
+	) -> color_eyre::Result<()> {
+		let futures = self.consumers.into_iter().map(|update_channels| {
+			Self::run_per_consumer(
+				update_channels,
+				self.urls.clone(),
+				self.from_block_number,
+				self.to_block_number,
+				shutdown_tx.clone(),
+				self.retry.clone(),
+			)
+		});
+
+		let mut flat_futures = futures.flatten().collect::<Vec<_>>();
+		flat_futures.extend(tasks);
+
+		tokio::select! {
+			_ = shutdown_future => {
+				info!("Shutting down chain head subscription on termination signal");
+			}
+			_ = future::try_join_all(flat_futures) => {
+				info!("Chain head subscription finished");
+			}
+		}
+
+		Ok(())
+	}
+}
+
+impl HistoricalSubscription {
+	pub fn new(
+		urls: Vec<String>,
+		from_block_number: BlockNumber,
+		to_block_number: BlockNumber,
+		retry: RetryOptions,
+	) -> HistoricalSubscription {
+		HistoricalSubscription { urls, from_block_number, to_block_number, consumers: Vec::new(), retry }
+	}
+
+	// Per consumer
+	async fn run_per_node(
+		mut update_channel: Sender<ChainSubscriptionEvent>,
+		url: String, // `String` rather than `&str` because we spawn this method as an asynchronous task
+		from_block_number: BlockNumber,
+		to_block_number: BlockNumber,
+		shutdown_tx: BroadcastSender<()>,
+		retry: RetryOptions,
+	) {
+		let mut shutdown_rx = shutdown_tx.subscribe();
+		let mut executor = RequestExecutor::new(retry);
+		const HEARTBEAT_INTERVAL: Duration = Duration::from_millis(200);
+		let mut heartbeat_periodic = interval_at(tokio::time::Instant::now() + HEARTBEAT_INTERVAL, HEARTBEAT_INTERVAL);
+
+		for block_number in from_block_number..=to_block_number {
```

Comment:
How are we handling `to_block_number` being actually greater than the tip of the chain ?

Reply:
good question

Reply:
now we're checking it after the tool connects to the chain.